File: examples/transactions/v2/go/offlineTransactionAuthorization.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io/ioutil"

	"crypto/ed25519"

	"github.com/algorand/go-algorand-sdk/client/v2/algod"
	"github.com/algorand/go-algorand-sdk/crypto"
	"github.com/algorand/go-algorand-sdk/encoding/msgpack"
	"github.com/algorand/go-algorand-sdk/mnemonic"
	"github.com/algorand/go-algorand-sdk/transaction"
	"github.com/algorand/go-algorand-sdk/types"
)

// user declared mnemonic for myAccount
const myMnemonic = "boy kidney fall hamster ecology mercy inquiry vast deal normal vibrant labor couch economy embody glory possible color burger addict soap almost margin about negative" // TODO:"Your 25-word mnemonic goes here";"
const receiver = "GD64YIY3TWGDMCNPP553DZPPR6LDUSFQOIJVFDPPXWEG3FVOJCCDBBHU5A"

// user declared algod connection parameters
const algodAddress = "http://localhost:49392"                                         //TODO:"http://localhost:4001"
const algodToken = "a31f09a18dbf7ad68c9e0ff22355774fb89c67ed2c4642d6c6822f9360cd7697" //TODO:"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";

// Function that waits for a given txId to be confirmed by the network
func waitForConfirmation(txID string, client *algod.Client) {
	status, err := client.Status().Do(context.Background())
	if err != nil {
		fmt.Printf("error getting algod status: %s\n", err)
		return
	}
	lastRound := status.LastRound
	for {
		pt, _, err := client.PendingTransactionInformation(txID).Do(context.Background())
		if err != nil {
			fmt.Printf("error getting pending transaction: %s\n", err)
			return
		}
		if pt.ConfirmedRound > 0 {
			fmt.Printf("Transaction "+txID+" confirmed in round %d\n", pt.ConfirmedRound)
			break
		}
		fmt.Printf("...waiting for confirmation\n")
		lastRound++
		status, err = client.StatusAfterBlock(lastRound).Do(context.Background())
	}
}

// utility function to get Account object
func getAccount(mn string) (string, ed25519.PrivateKey) {
	sk, err := mnemonic.ToPrivateKey(mn)
	if err != nil {
		fmt.Printf("...error recovering account: %s\n", err)
	}
	pk := sk.Public()
	var a types.Address
	cpk := pk.(ed25519.PublicKey)
	copy(a[:], cpk[:])
	fmt.Printf("...found address: %s\n", a.String())
	address := a.String()
	return address, sk
}

func createTransaction(algodClient *algod.Client, myAccount string) types.Transaction {
	// get network suggested parameters
	txParams, err := algodClient.SuggestedParams().Do(context.Background())
	if err != nil {
		fmt.Printf("...error getting suggested tx params: %s\n", err)
	}
	txParams.FlatFee = true
	var minFee uint64 = 1000
	genID := txParams.GenesisID
	genHash := txParams.GenesisHash
	firstValidRound := uint64(txParams.FirstRoundValid)
	lastValidRound := uint64(txParams.LastRoundValid)

	// make transaction
	fmt.Println("Creating transaction...")
	// from account to receiver
	fromAddr := myAccount
	toAddr := receiver
	var amount uint64 = 1000000

	txnObj, err := transaction.MakePaymentTxnWithFlatFee(fromAddr, toAddr, minFee, amount, firstValidRound, lastValidRound, nil, "", genID, genHash)
	if err != nil {
		fmt.Printf("...error creating transaction: %s\n", err)
	}
	fmt.Printf("...tx1: from %s to %s for %v microAlgos\n", fromAddr, toAddr, amount)

	return txnObj
}
func saveUnsignedTransactionToFile(txnObj types.Transaction) {
	// assign Transaction object to SignedTxn struct
	unsignedTxn := types.SignedTxn{
		Txn: txnObj,
	}

	// save unsigned Transaction to file
	err := ioutil.WriteFile("./unsigned.txn", msgpack.Encode(unsignedTxn), 0644)
	if err == nil {
		fmt.Printf("Saving unsigned transaction to file...\n")
		return
	}
	fmt.Printf("...failed to save transaction to file, error %s\n", err)
}

func readUnsigedTransactionFromFile() types.Transaction {
	// read unsigned transaction from file
	fmt.Println("Reading transaction from file...")
	bytesRead, err := ioutil.ReadFile("./unsigned.txn")
	if err != nil {
		fmt.Printf("...error reading transaction from file: %s\n", err)
	}
	fmt.Println("Decoding file bytes...")
	var unsignedTxn types.SignedTxn
	msgpack.Decode(bytesRead, &unsignedTxn)
	txnObj := unsignedTxn.Txn

	// TODO: display unsigned transaction

	return txnObj
}

func signTransaction(txnObj types.Transaction, sk ed25519.PrivateKey) []byte {
	// sign the transaction
	fmt.Println("Signing transactions...")
	signedTxnID, signedBytes, err := crypto.SignTransaction(sk, txnObj)
	if err != nil {
		fmt.Printf("...Failed to sign transaction: %s\n", err)
	}
	fmt.Println("...Signed transaction: ", signedTxnID)

	return signedBytes
}

func saveSignedTransactionToFile(signedBytes []byte) {
	// save the signed transaction to file
	fmt.Println("Saving signed transction to file...")
	err := ioutil.WriteFile("./signed.stxn", signedBytes, 0644)
	if err == nil {
		fmt.Printf("...Saved signed transaction to file\n")
		return
	}
	fmt.Printf("...Failed to save transaction to file, error %s\n", err)
}

func readSignedTransactionFromFile() []byte {
	// read unsigned transaction from file
	fmt.Println("Reading transaction from file...")
	signedBytes, err := ioutil.ReadFile("./signed.stxn")
	if err != nil {
		fmt.Printf("...Error reading signed transaction from file: %s\n", err)
	}
	fmt.Println("Decoding bytes...")

	// display signed transaction
	var signedTxn types.SignedTxn
	var signedTxnObj types.Transaction
	msgpack.Decode(signedBytes, &signedTxn)
	signedTxnObj = signedTxn.Txn

	txnJSON, err := json.MarshalIndent(signedTxnObj, "", "\t")
	if err != nil {
		fmt.Printf("Can not marshall txn data: %s\n", err)
	}
	fmt.Printf("Transaction information: %s\n", txnJSON)

	return signedBytes
}

func sendSignedTransaction(algodClient *algod.Client, signedBytes []byte) {
	// send the transaction to the network
	fmt.Println("Sending signed transction to network...")
	txID, err := algodClient.SendRawTransaction(signedBytes).Do(context.Background())
	if err != nil {
		fmt.Printf("...Failed to send transaction: %s\n", err)
		return
	}

	// wait for response
	waitForConfirmation(txID, algodClient)
}

func main() {
	// Initialize an algodClient
	algodClient, err := algod.MakeClient(algodAddress, algodToken)
	if err != nil {
		fmt.Printf("failed to make algod client: %v\n", err)
		return
	}

	// Load account from Mymnemonic
	fmt.Println("Loading signing account...")
	address, sk := getAccount(myMnemonic)

	// Create transaction object from account
	txnObj := createTransaction(algodClient, address)

	// Save unsigned transaction to file
	saveUnsignedTransactionToFile(txnObj)

	// Read the unsigned transaction from the file
	unsignedTxn := readUnsigedTransactionFromFile()

	// Sign the transaction using the mnemonic
	signedBytes := signTransaction(unsignedTxn, sk)

	// Save the signed transaction to file
	saveSignedTransactionToFile(signedBytes)

	// Read the signed transaction from file
	signedBytes = readSignedTransactionFromFile()

	// Send the transaction to the network
	sendSignedTransaction(algodClient, signedBytes)
}

File: examples/transactions/v2/python/offline_transaction_authorization.py
import json
import time
import base64
import os
from algosdk.v2client import algod
from algosdk import mnemonic
from algosdk.future import transaction
from algosdk import encoding
from algosdk import account

# user declared mnemonic for myAccount
my_mnemonic = "boy kidney fall hamster ecology mercy inquiry vast deal normal vibrant labor couch economy embody glory possible color burger addict soap almost margin about negative" # TODO:"Your 25-word mnemonic goes here";"
receiver = "GD64YIY3TWGDMCNPP553DZPPR6LDUSFQOIJVFDPPXWEG3FVOJCCDBBHU5A"

# user declared algod connection parameters
algod_address = "http://localhost:49392"                                          #TODO:"http:#localhost:4001"
algod_token = "a31f09a18dbf7ad68c9e0ff22355774fb89c67ed2c4642d6c6822f9360cd7697" #TODO:"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";

# Function that waits for a given txid to be confirmed by the network
def wait_for_confirmation(client, txid):
    last_round = client.status().get('last-round')
    txinfo = client.pending_transaction_info(txid)
    while not (txinfo.get('confirmed-round') and txinfo.get('confirmed-round') > 0):
        print("Waiting for confirmation...")
        last_round += 1
        client.status_after_block(last_round)
        txinfo = client.pending_transaction_info(txid)
    print("Transaction {} confirmed in round {}.".format(txid, txinfo.get('confirmed-round')))
    return txinfo

# utility def to get address string
def get_account(mn) :
    print("Loading signing account...")
    sk = mnemonic.to_private_key(mn)
    address = account.address_from_private_key(sk)
    print("...found address : ", address)
    return address, sk

def create_transaction(algod_client, address) :
    print("Creating transaction...")
	# get node suggested parameters
    params = algod_client.suggested_params()
    # comment out the next two (2) lines to use suggested fees
    params.flat_fee = True
    params.fee = 1000
    sender = address
    amount = 1000000
    txn_obj = transaction.PaymentTxn(sender, params, receiver, amount)
    print("...txn: from {} to {} for {} microAlgos".format(sender, receiver, amount))
    print("...with txid:", txn_obj.get_txid())

    return txn_obj

def save_unsigned_transaction_to_file(txn_obj) :
    print("Saving unsigned transction to file...")
    dir_path = os.path.dirname(os.path.realpath(__file__))
    transaction.write_to_file([txn_obj], dir_path + "/unsigned.txn")

def read_unsiged_transaction_from_file() :
    print("Reading unsigned transction from file...")
    dir_path = os.path.dirname(os.path.realpath(__file__))
    txns = transaction.retrieve_from_file(dir_path + "/unsigned.txn")
    unsigned_txn = txns[0]

    return unsigned_txn

def sign_transaction(unsigned_txn, sk) :
    print("Signing transaction...")
    signed_txn = unsigned_txn.sign(sk)
    txid = signed_txn.transaction.get_txid()
    print("Signed transaction with txID: {}".format(txid))

    return signed_txn

def save_signed_transaction_to_file(signed_txn) :
    print("Saving signed transction to file...")
    dir_path = os.path.dirname(os.path.realpath(__file__))
    transaction.write_to_file([signed_txn], dir_path + "/signed.txn")

def read_signed_transaction_from_file() :
    print("Reading signed transction from file...")
    dir_path = os.path.dirname(os.path.realpath(__file__))
    txns = transaction.retrieve_from_file(dir_path + "/signed.txn")
    signed_txn = txns[0]

    #TODO: Ensure signed_txn contains a signature. See issue: https://github.com/algorand/py-algorand-sdk/issues/124

    return signed_txn

def send_signed_transaction(algod_client, signed_txn) :
    # send transactions
    print("Sending transactions...")
    tx_id = algod_client.send_transactions([signed_txn])

    # wait for confirmation
    wait_for_confirmation(algod_client, tx_id) 

    confirmed_txn = algod_client.pending_transaction_info(tx_id.get_txid())
    print("Transaction information: {}".format(json.dumps(confirmed_txn, indent=4)))

def main() :
	# Initialize an algod_client
    algod_client = algod.AlgodClient(algod_token, algod_address)

	# Load account from Mymnemonic
    address, sk = get_account(my_mnemonic)

	# Create transaction object from account
    txn_obj = create_transaction(algod_client, address)

	# Save unsigned transaction to file
    save_unsigned_transaction_to_file(txn_obj)

	# Read the unsigned transaction from the file
    unsigned_txn = read_unsiged_transaction_from_file()

	# Sign the transaction using the mnemonic
    signed_txn = sign_transaction(unsigned_txn, sk)

	# Save the signed transaction to file
    save_signed_transaction_to_file(signed_txn)

	# Read the signed transaction from file
    signed_txn = read_signed_transaction_from_file()

	# Send the transaction to the network
    send_signed_transaction(algod_client, signed_txn)

main()